### ISSUE TITLE
Payday 2 Linux is missing DB:create_entry

### ISSUE THREAD

**Issue by zekesonxx** [blt4linux/blt4l#70]
From trying to use the [Megalovania](https://modworkshop.net/mydownloads.php?action=view_down&did=18637) mod with BeardLib and the [BeardLib Music Module](https://modworkshop.net/mydownloads.php?action=view_down&did=18107):

```
09:14:05 PM Error: mods/BeardLib/BeardLibCore.lua:197: attempt to call method 'create_entry' (a nil value)
09:14:05 PM Lua: [Megalovania] [ERROR] An error occured on initilization of module: AddFiles. Error:
nil
```

Is this something that could be implemented in blt4l, similar to the mod_overrides fixes?

**Comment by RomanHargrave** [comment]
This is something I have been working on. Unfortunately, it seems like a lot of the code for interacting with the asset database was inlined during compilation, which makes it that much harder to reverse engineer.

I can give you some more in depth information if you want to have a go at REing the necessary parts so that we can retrofit support.

**Comment by RomanHargrave** [comment]
@zekesonxx Maybe we should start gathering as much information about how `dsl::MainDB` works in order to figure out how we could add features to it.

**Comment by punkkeks** [comment]
Can you please explain, why it just works in the windows version of blt? I dont find any clue implementing it in the original github repo.

**Comment by zekesonxx** [comment]
@punkkeks BLT isn't responsible for adding this function, it's part of the base game.

Except the Linux version is missing it, because Overkill is incompetent.

**Comment by punkkeks** [comment]
Can't we just hook that and create our own db? 😁

Am 7. September 2017 17:42:52 MESZ schrieb Zeke Sonxx <notifications@github.com>:

**Comment by zekesonxx** [comment]
Kind of. That's what this issue is about.

The function is responsible for adding content (models/textures/audio/etc) to the Diesel engine's database. Reimplementing the function requires us to do the same, we can't just make a fake database.

**Comment by RomanHargrave** [comment]
Not exactly. DB in lua refers to the asset database, so for it to work, PAYDAY has to know about changes somehow.

---- On Thu, 07 Sep 2017 08:44:11 -0700 notifications@github.com wrote ----

Can't we just hook that and create our own db? 😁

Am 7. September 2017 17:42:52 MESZ schrieb Zeke Sonxx <notifications@github.com>:

**Comment by Fanged-Hex** [comment]
One of VoidUI's developer here.
So is there anything new about this issue ? We will really love to support Linux with our mod but we use custom textures :(

**Comment by ZNixian** [comment]
@Fanged-Hex Unfortunately, the only current option is to somehow avoid using `entity_create`. Unless you found some random texture that's not being used and `mod_override`ed it, there's not much you can do.

I've been on-and-off doing some work pulling apart the internals of PAYDAY 2, very slowly building our own header files. Hopefully, this will eventually be enough for BLT to add this method. This is being slowed down quite a lot because of inlining (if you call a method, the compiler sometimes copies the contents of that method to where you called it, making the game faster), which makes it a lot harder to reverse engineer.

**Comment by Fanged-Hex** [comment]
Don't worry, I was just asking :P 
Thanks anyway for those explanation, so as I heard, mod_overrides works now on Linux ?

**Comment by zekesonxx** [comment]
mod_overrides now works on Linux without blt4l, yes.

**Comment by dribbleondo** [comment]
The fact that Overkill never told anyone that mod-overrides now works natively is a sign on bad communication on their part.

**Comment by Fanged-Hex** [comment]
I know that Overkill are "skilled" developers and know what they are doing lol

**Comment by Fanged-Hex** [comment]
About what @dribbleondo said, I can confirm their lack of com, cause I sent a mail to them like 1 month ago, they didn't opened it yet ...

**Comment by ZNixian** [comment]
IMO the best way of achieving this would be to find all the corresponding functions between the GNU+Linux and Windows binaries, probably work out some of the offsets for the most critical classes (on both platforms) then try and write duplicates of the originals.

**Comment by ZNixian** [comment]
@RomanHargrave I think I've found how mod_overrides is implemented by the engine.

By hooking all four varients of `dsl::DB::do_resolve`, we should be able to create and return a custom asset. This function returns a `dsl::Archive` instance.

`dsl::MainDB::_get_mod_override` is a good example of how to create an archive from a path, particularly the `dsl::FileSystemStack::open` call.

**Comment by dribbleondo** [comment]
If this turns out to be correct then it feels like we have found the master key to unlocking Payday 2's full potential.

**Comment by ZNixian** [comment]
I've been able to load the game successfully using a custom implementation of the function, so it most certainly should work.

**Comment by ZNixian** [comment]
I've been able to get a prototype of this working, main thing now is cleaning stuff up and writing the Lua interface.

**Comment by ZNixian** [comment]
@Fanged-Hex I've got a basic version that works fine with VoidUI, I'll clean it up and submit a PR later today.

**Comment by ZNixian** [comment]
See #100.

**Comment by Fanged-Hex** [comment]
Nice work :)
I will look into it when I got time :D

**Comment by ZNixian** [comment]
Just a heads-up that you have to have a bunch of stuff installed to compile it, if you're compiling it yourself.

**Comment by dribbleondo** [comment]
I'm gonna be helpful and compile this on my debian (Linux Mint) distro in a few hours, that'll make it easier and a bit more consistent for testing purposes.

**Comment by dribbleondo** [comment]
I tried compiling the latest version of the loader, but it seems VoidUI doesn't want to present itself :/

**Comment by ZNixian** [comment]
@dribbleondo Did you use `-DLIBCXX=1`? I probably should have put that in some kind of documentation.

**Comment by dribbleondo** [comment]
I tried compiling with that variable, but I get thrown an error saying that cmake won't use the variable.

**Comment by ZNixian** [comment]
You need LibCXX and Clang++ installed.

**Comment by dribbleondo** [comment]
I installed libCXX and Clang 3.8 and It just remained the same. I'm probably at fault, I just don't know why =/

If it helps, **[I uploaded a Pastebin](https://pastebin.com/ztiDkULi)** of my terminal output.

**Comment by hellow554** [comment]
It's called `-DUSE_LIBCXX=1`

Am 10. Januar 2018 12:47:49 MEZ schrieb dribbleondo <notifications@github.com>:

**Comment by ZNixian** [comment]
My bad, it's `-DUSE_LIBCXX=1`

**Comment by dribbleondo** [comment]
I can confirm using that command in front of Cmake solved the issue. Had to install libc++dev as well, but now VoidUI works!

`cmake -DUSE_LIBCXX=1 ..`

**Comment by ZNixian** [comment]
Closed by #100.
